const-eval.const-context.outer-generics: make more clear/obvious (#2159)
rust-lang/reference: src/const_eval.md

* A [const block]

r[const-eval.const-context.outer-generics]

> Const contexts that are used as parts of types (array type expressions, array repeat length expressions, and const generic arguments) can only make restricted use of surrounding generic parameters: such an expression must either be a single bare const generic parameter, or an arbitrary expression not making use of any generics.

Reviewer: I'm struggling a bit with how this is worded. The "used as parts of types" seems at odds with the parenthetical that says "array repeat length expressions", since that isn't a type. For example:

```rust
fn d<const N: usize>() {
    // Use of bare N in type and repeat expr is OK.
    let x: [i32; N] = [1; N];
    let y: _ = [1; N+1]; // ERROR: N may only be used as standalone arguments
}
```

How about we simplify it a little and turn it around:

```suggestion
Array type length expressions, array repeat length expressions, and const generic arguments are restricted in their use of outer generic parameters: such an expression must either be a single const generic parameter, or an expression that does not reference any generic parameters.
```
